File: prisma/refs/design-twitter-feed.ts
export class Twitter {
  private time = 0;
  private tweets = new Map<number, [number, number][]>(); // user -> [time, tweetId]
  private follows = new Map<number, Set<number>>();

  postTweet(userId: number, tweetId: number): void {
    if (!this.tweets.has(userId)) this.tweets.set(userId, []);
    this.tweets.get(userId)!.push([this.time++, tweetId]);
  }

  getNewsFeed(userId: number): number[] {
    const users = new Set(this.follows.get(userId) ?? []);
    users.add(userId);
    const all: [number, number][] = [];
    for (const u of users) all.push(...(this.tweets.get(u) ?? []));
    return all
      .sort((a, b) => b[0] - a[0])
      .slice(0, 10)
      .map(([, id]) => id);
  }

  follow(followerId: number, followeeId: number): void {
    if (!this.follows.has(followerId)) this.follows.set(followerId, new Set());
    this.follows.get(followerId)!.add(followeeId);
  }

  unfollow(followerId: number, followeeId: number): void {
    this.follows.get(followerId)?.delete(followeeId);
  }
}

File: prisma/refs/java/design-twitter-feed.java
import java.util.*;

class Twitter {
    private int time = 0;
    private final Map<Integer, List<int[]>> tweets = new HashMap<>();  // user -> [time, tweetId]
    private final Map<Integer, Set<Integer>> follows = new HashMap<>();

    public Twitter() {}

    public void postTweet(int userId, int tweetId) {
        tweets.computeIfAbsent(userId, k -> new ArrayList<>()).add(new int[]{time++, tweetId});
    }

    public List<Integer> getNewsFeed(int userId) {
        Set<Integer> users = new HashSet<>(follows.getOrDefault(userId, Collections.emptySet()));
        users.add(userId);
        List<int[]> all = new ArrayList<>();
        for (int u : users) all.addAll(tweets.getOrDefault(u, Collections.emptyList()));
        all.sort((a, b) -> b[0] - a[0]);
        List<Integer> out = new ArrayList<>();
        for (int i = 0; i < Math.min(10, all.size()); i++) out.add(all.get(i)[1]);
        return out;
    }

    public void follow(int followerId, int followeeId) {
        follows.computeIfAbsent(followerId, k -> new HashSet<>()).add(followeeId);
    }

    public void unfollow(int followerId, int followeeId) {
        Set<Integer> s = follows.get(followerId);
        if (s != null) s.remove(followeeId);
    }
}
